src/components/Plugin/Plugin.js

```
@@ -90,24 +98,22 @@ export class Plugin extends Component {
     }
   }
 
-  fetchPluginData() {
+  async fetchPluginData() {
     const { plugin: pluginId } = this.props.match.params;
 
     let pluginData;
-    return new Promise(async (resolve, reject) => {
-      try {
-        const plugin = await this.client.getPlugin(parseInt(pluginId, 10));
-        pluginData = plugin.data;
-        pluginData.url = plugin.url;
-      } catch (e) {
-        return reject(e);
-      }
+    try {
+      const plugin = await this.client.getPlugin(parseInt(pluginId, 10));
+      pluginData = plugin.data;
+      pluginData.url = plugin.url;
+    } catch (e) {
+      return Promise.reject(e);
```

I think you should just not bother catching here if the exception cannot be handled by this function.

I am not sure it makes sense to `return Promise.reject(e)`, especially not in an `async` block

---

Returning a rejected Promise here is so that the catch error can be "bubbled up" to the componentDidMount where this function is called and then it can be handled by the catch block there. This way all errors can be handled in the same place (i.e. the catch block in the CDM method)

https://github.com/FNNDSC/ChRIS_store_ui/blob/63d9b5d84e1620048721abe54d059c42b41b2f02/src/components/Plugin/Plugin.js#L48-L52

If this seems like an anti-pattern then the errors can be handled where they happen.

---

`return` _typically_ implies that the state is expected and that the data is sane. The safer way of handling errors would be to let exceptions be thrown and caught. To handle this exception, wouldn't you need something like

```javascript
let result = await fun();
if (result instanceof Promise.reject) ...
```

---

In other words, exceptions mean "something unexpected happened, and the app's state (data) doesn't make sense anymore." Thus when a function does not throw an exception, it means "everything worked as expected." If you catch an exception without being able to restore the state (in this case, set `loading: false` and get stars) then the component won't know that something went wrong, and it'll keep trying to work on bad data (an unexpected `Promise.reject` object) leading to more data/state corruption.

Catch an exception if:

- the state can be corrected somehow (e.g. ok, something went wrong, let's show an error message and stop making things worse)
- within the function's scope it can handle the exception in a way that is better than how its parent might handle the exception
- we can enhance the exception before rethrowing it with extra information from the current scope

---

Alright I understand. It was wrong to throw the same Exception twice. I've now switched to handling the errors where they happen, if they can be, otherwise let the function reject the Promise.